[15_0_X] Use HardwareResourcesDescription in ProcessConfiguration

makortel opened the pull request:
#### PR description:

Backport of https://github.com/cms-sw/cmssw/pull/47280, https://github.com/cms-sw/cmssw/pull/47355,  https://github.com/cms-sw/cmssw/pull/47473 (well, backported by dropping out the commit reverted there), and https://github.com/cms-sw/cmssw/pull/47477 (the commit of the PR split in two, one to replace the commit dropped in the earlier step, and the rest included in the last commit of this PR).

Resolves https://github.com/cms-sw/framework-team/issues/1248

#### PR validation:

Code compiles (plus the tests in https://github.com/cms-sw/cmssw/pull/47280 and https://github.com/cms-sw/cmssw/pull/47355)

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Backport of https://github.com/cms-sw/cmssw/pull/47280 and https://github.com/cms-sw/cmssw/pull/47355

makortel committed:
Move ResourceInformation abstract base class to FWCore/AbstractServices

makortel committed:
Change ResourceInformation::acceleratorTypes() to be a vector<string>

Just storing the strings from process.options.accelerators is more
general as it avoids updating the ResourceInformation(Service) every
time a new accelerator type is added to the system (outside of the
framework).

makortel committed:
Add hasGpuNvidia() function to ResourceInformation service

makortel committed:
Use ResourceInformation::hasGpuNvidia() in TensorFlow

makortel committed:
Add a function in ResourceInformation to return HardwareResourcesDescription

makortel committed:
Use HardwareResourcesDescription in ProcessConfiguration

makortel committed:
Keep ResourceInformation service optional in ScheduleItems

makortel committed:
Extend Process{Configration,History}_t tests to cover HardwareResourcesDescription

makortel committed:
Rename TestIOPool{Input,Streamer}ReducedProcessHistory tests

makortel committed:
Add tests for merging ROOT and streamer files with different hardware resources

makortel committed:
Make empty HardwareResourcesDescription to serialize as an empty string

cmsbuild commented:
A new Pull Request was created by  @makortel for CMSSW_15_0_X.

It involves the following packages:

- DQM/SiStripMonitorHardware (**dqm**)
- DQMServices/FwkIO (**dqm**)
- DataFormats/Provenance (**core**)
- FWCore/AbstractServices (**core**)
- FWCore/Framework (**core**)
- FWCore/Integration (**core**)
- FWCore/Services (**core**)
- FWCore/Sources (**core**)
- FWCore/TestProcessor (**core**)
- FWCore/Utilities (**core**)
- GeneratorInterface/LHEInterface (**generators**)
- HeterogeneousCore/CUDAServices (**heterogeneous**)
- HeterogeneousCore/ROCmServices (**heterogeneous**)
- IOPool/Common (**core**)
- IOPool/Input (**core**)
- IOPool/SecondaryInput (**core**)
- IOPool/Streamer (**core**)
- Mixing/Base (**simulation**)
- PhysicsTools/PyTorch (**ml**)
- PhysicsTools/TensorFlow (**ml**)


@Dr15Jones, @antoniovagnerini, @bbilin, @civanch, @cmsbuild, @fwyzard, @kpedro88, @lviliani, @makortel, @mdhildreth, @menglu21, @mkirsano, @rseidita, @smuzaffar, @valsdav, @y19y19 can you please review it and eventually sign? Thanks.
@alberto-sanchez, @arossi83, @barvic, @fabiocos, @felicepantaleo, @fioriNTU, @fwyzard, @idebruyn, @jandrea, @missirol, @mkirsano, @mmusich, @richa2710, @riga, @rovere, @sroychow, @threus, @wddgit this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"052ebbdd12cc4a9905fb8a3ef801885075fe1158":{"files":["DataFormats/Provenance/interface/HardwareResourcesDescription.h","DataFormats/Provenance/src/HardwareResourcesDescription.cc","DataFormats/Provenance/test/HardwareResourcesDescription_t.cpp"],"squashed":false,"time":1740058411},"07ddd0b0b527d9b0333fc0cde786ddaf3c50782b":{"files":["FWCore/Integration/test/ProcessConfiguration_t.cpp","FWCore/Integration/test/ProcessHistory_t.cpp"],"squashed":false,"time":1740058464},"2f5c577ff75de2ca1bde79fc3345c72f003d55bb":{"files":["IOPool/Input/test/BuildFile.xml","IOPool/Input/test/testReducedProcessHistoryVersion.sh","IOPool/Streamer/test/BuildFile.xml","IOPool/Streamer/test/run_TestReducedProcessHistoryVersion.sh"],"squashed":false,"time":1740058475},"3b590e2cbc653f6ac18bf58a6a50eb629e337ad7":{"files":["DQM/SiStripMonitorHardware/src/SiStripSpyEventMatcher.cc","DQMServices/FwkIO/plugins/DQMRootOutputModule.cc","DQMServices/FwkIO/plugins/format.h","DataFormats/Provenance/interface/ModuleDescription.h","DataFormats/Provenance/interface/ProcessConfiguration.h","DataFormats/Provenance/src/ModuleDescription.cc","DataFormats/Provenance/src/ProcessConfiguration.cc","DataFormats/Provenance/src/ProcessHistory.cc","FWCore/Framework/src/ScheduleItems.cc","FWCore/Framework/test/Event_t.cpp","FWCore/Framework/test/checkForModuleDependencyCorrectness_t.cppunit.cc","FWCore/Framework/test/edproducer_productregistry_callback.cc","FWCore/Framework/test/event_getrefbeforeput_t.cppunit.cc","FWCore/Framework/test/eventprincipal_t.cppunit.cc","FWCore/Framework/test/generichandle_t.cppunit.cc","FWCore/Framework/test/makeDummyProcessConfiguration.h","FWCore/Framework/test/maker2_t.cppunit.cc","FWCore/Integration/plugins/DelayedReaderThrowingSource.cc","FWCore/Integration/plugins/PutOrMergeTestSource.cc","FWCore/Integration/test/ProcessConfiguration_t.cpp","FWCore/Integration/test/ProcessHistory_t.cpp","FWCore/Sources/src/DaqProvenanceHelper.cc","FWCore/TestProcessor/src/TestProcessor.cc","GeneratorInterface/LHEInterface/plugins/BuildFile.xml","GeneratorInterface/LHEInterface/plugins/LHEProvenanceHelper.cc","IOPool/Common/bin/EdmProvDump.cc","IOPool/Common/test/removeChangingParts.py","IOPool/Common/test/unit_test_outputs/provdump.log","IOPool/Input/src/ProvenanceAdaptor.cc","IOPool/SecondaryInput/test/SecondaryProducer.cc","Mixing/Base/src/PileUp.cc"],"squashed":false,"time":1740058445},"491cfcae9b8bc761ebb0699368129ff89cd78183":{"files":["FWCore/AbstractServices/interface/ResourceInformation.h","FWCore/Framework/src/ScheduleItems.cc","FWCore/Services/plugins/ResourceInformationService.cc","FWCore/Services/test/test_resourceInformationService.sh"],"squashed":false,"time":1740058366},"4a9711e7833d0f6ad0c2b07d899176b559dc30cc":{"files":["FWCore/AbstractServices/BuildFile.xml","FWCore/AbstractServices/interface/ResourceInformation.h","FWCore/AbstractServices/src/ResourceInformation.cc","FWCore/Framework/BuildFile.xml","FWCore/Framework/src/ScheduleItems.cc","FWCore/Services/BuildFile.xml","FWCore/Services/plugins/CPU.cc","FWCore/Services/plugins/CondorStatusUpdater.cc","FWCore/Services/plugins/ResourceInformationService.cc","FWCore/Utilities/src/ResourceInformation.cc","HeterogeneousCore/CUDAServices/plugins/BuildFile.xml","HeterogeneousCore/CUDAServices/plugins/CUDAService.cc","HeterogeneousCore/CUDAServices/test/BuildFile.xml","HeterogeneousCore/CUDAServices/test/testCUDAService.cpp","HeterogeneousCore/ROCmServices/plugins/BuildFile.xml","HeterogeneousCore/ROCmServices/plugins/ROCmService.cc","HeterogeneousCore/ROCmServices/test/BuildFile.xml","HeterogeneousCore/ROCmServices/test/testROCmService.cpp","PhysicsTools/PyTorch/test/BuildFile.xml","PhysicsTools/PyTorch/test/testBaseCUDA.h","PhysicsTools/TensorFlow/BuildFile.xml","PhysicsTools/TensorFlow/src/TensorFlow.cc","PhysicsTools/TensorFlow/test/BuildFile.xml","PhysicsTools/TensorFlow/test/testBaseCUDA.h"],"squashed":false,"time":1740058350},"5599cf26a76a52b438d7f624cc4466d91264c70a":{"files":["FWCore/AbstractServices/BuildFile.xml","FWCore/AbstractServices/interface/ResourceInformation.h","FWCore/Services/plugins/ResourceInformationService.cc"],"squashed":false,"time":1740058406},"6cd0a04a7e3705e006487e83af8a888899091ff7":{"files":["IOPool/Common/bin/EdmProvDump.cc"],"squashed":false,"time":1740058469},"af07337994c6156762fbcda24d1cfd6b2dac5bef":{"files":["FWCore/AbstractServices/interface/ResourceInformation.h","FWCore/Services/plugins/ResourceInformationService.cc","HeterogeneousCore/CUDAServices/test/testCUDAService.cpp"],"squashed":false,"time":1740058370},"c8326daa2b7c8c8e2574e5b4bea1af1a035c53ff":{"files":["IOPool/Input/test/BuildFile.xml","IOPool/Input/test/testReducedProcessHistoryCreate_cfg.py","IOPool/Input/test/testReducedProcessHistoryHardwareResources.sh","IOPool/Streamer/test/BuildFile.xml","IOPool/Streamer/test/run_TestReducedProcessHistoryHardwareResources.sh","IOPool/Streamer/test/testReducedProcessHistoryCreate_cfg.py"],"squashed":false,"time":1740058480},"dd9ed4f3c3d349d89fea06adbc3755c94d5ad854":{"files":["FWCore/Framework/src/ScheduleItems.cc"],"squashed":false,"time":1740058458},"ef280e0132ff2629c98c5167da44018875442cb5":{"files":["PhysicsTools/TensorFlow/src/TensorFlow.cc"],"squashed":false,"time":1740058376}},"emoji":{},"last_seen_sha":"c8326daa2b7c8c8e2574e5b4bea1af1a035c53ff","signatures":{}} -->

makortel commented:
enable gpu

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 220KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9d1efb/44545/summary.html
**COMMIT**: c8326daa2b7c8c8e2574e5b4bea1af1a035c53ff
**CMSSW**: CMSSW_15_0_X_2025-02-20-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47416/44545/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-20-1100+9d1efb/67042/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 9 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-20-1100+9d1efb/67042/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4018889
 - DQMHistoTests: [Total failures: 68](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-20-1100+9d1efb/67042/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4018801
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 218 log files, 189 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-20-1100+9d1efb/67042/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 24 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-02-20-1100+9d1efb/67041/validateJR.html)
 - DQMHistoTests: Total files compared: 7
 - DQMHistoTests: Total histograms compared: 53071
 - DQMHistoTests: [Total failures: 869](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-02-20-1100+9d1efb/67041/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 52202
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 6 files compared)
 - Checked 24 log files, 30 edm output root files, 7 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-02-20-1100+9d1efb/67041/triggerResults): no differences found

makortel commented:
CPU comparison differences are related to https://github.com/cms-sw/cmssw/issues/47071

GPU comparison differences look compatible with the non-reproducibilities in the pixel code

makortel commented:
+core

makortel commented:
+heterogeneous

antoniovagnerini commented:
+dqm
- spurious differences from unstable WF `29834.999`(open issue #47071)

lviliani commented:
+1

makortel commented:
@cms-sw/ml-l2 Could you review and sign, please? I'd like to get this PR in 15_0_0 if possible. Thanks!

For @cms-sw/orp-l2 Despite of me wanting to get this PR in 15_0_0, I would not hold 15_0_0 for this PR alone. It can go in also later.

makortel commented:
@cms-sw/ml-l2 ping

cmsbuild commented:
Pull request has been put on hold by  @makortel
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

cmsbuild commented:
Pull request #47416 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar, @valsdav, @y19y19 can you please check and sign again.

makortel commented:
I took out the commit that altered the `edmProvDump` output. I opened https://github.com/cms-sw/cmssw/pull/47473 to revert the corresponding commit from `master` so we can check that it is sufficient to get CRAB working again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests GpuUnitTests 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9d1efb/44723/summary.html
**COMMIT**: 29ae1f5287bbae0cea933da169a2932afeede6f2
**CMSSW**: CMSSW_15_0_X_2025-02-27-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47416/44723/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 2 errors in the following unit tests:

<pre>---> test TestIOPoolInputReducedProcessHistoryHardwareResources had ERRORS
---> test TestIOPoolStreamerReducedProcessHistoryHardwareResources had ERRORS
</pre>

## GPU Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test testTorchSimpleDnnCUDA had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-27-1100+9d1efb/67195/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 2 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-27-1100+9d1efb/67195/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4018889
 - DQMHistoTests: [Total failures: 123](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-27-1100+9d1efb/67195/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4018746
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 218 log files, 189 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-27-1100+9d1efb/67195/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 8 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-02-27-1100+9d1efb/67194/validateJR.html)
 - DQMHistoTests: Total files compared: 7
 - DQMHistoTests: Total histograms compared: 53071
 - DQMHistoTests: [Total failures: 380](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-02-27-1100+9d1efb/67194/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 52691
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 6 files compared)
 - Checked 24 log files, 30 edm output root files, 7 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-02-27-1100+9d1efb/67194/triggerResults): no differences found

makortel committed:
Add optional HardwareResourcesInformation printout to edmProvDump

For the moment CRAB needs the edmProvDump without the hardware
information, but two tests need the hardware information.

cmsbuild commented:
Pull request #47416 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar, @valsdav, @y19y19 can you please check and sign again.

makortel commented:
@cmsbuild, please test

makortel commented:
Ok, now this PR corresponds to the status in the `master` branch (with a bit different history though), and should work for both CRAB and the unit tests.

valsdav commented:
+ml

makortel commented:
CPU comparison differences are related to https://github.com/cms-sw/cmssw/issues/39803 and https://github.com/cms-sw/cmssw/issues/47071

GPU comparison differences seem compatible with pixel code non-reproducibilities, and https://github.com/cms-sw/cmssw/issues/47406

makortel commented:
Tests in 15_1_X IBs have not revealed any new problems, so this should be good to go now.

makortel commented:
+core

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_15_0_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_15_1_X is complete. This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @sextonkennedy, @rappoccio, @mandrenguyen (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
